apps/app_rpt.c

```diff
@@ -6761,12 +6796,45 @@ static int rpt_exec(struct ast_channel *chan, const char *data)
 			send_newkey(chan);
 			rpt_mutex_unlock(&myrpt->blocklock);
 		}
-		if (!strcasecmp(ast_channel_tech(l->chan)->type, "echolink") || !strcasecmp(ast_channel_tech(l->chan)->type, "tlb") || (l->name[0] > '9')) {
-			rpt_telemetry(myrpt, CONNECTED, l);
+
+		/* Be extra careful here.
+		 * Since above we impart this into the list of links (rpt_link_add),
+		 * it is possible that this node's thread (started at rpt() - the C function, not the dialplan one),
+		 * has already requested this channel be hungup.
+		 * We take care to not use ast_hangup if we detect the channel has a PBX,
+		 * so this channel won't have the rug pulled out from underneath it.
+		 * However, l->chan could still be set to NULL. Thus, if it's NULL here,
+		 * that means we're going to get hung up and should just abort, no point in doing telemetry.
+		 * For that reason, even though l->chan is chan if l->chan isn't NULL,
+		 * we always use chan here, rather than l->chan, since l->chan could be NULL.
+		 * This way, even if l->chan isn't NULL when we check but becomes NULL later, we're still safe. */
+		rpt_mutex_lock(&myrpt->lock);
+		if (l->chan) {
+			rpt_mutex_unlock(&myrpt->lock);
+			if (!strcasecmp(ast_channel_tech(chan)->type, "echolink") || !strcasecmp(ast_channel_tech(chan)->type, "tlb") || (l->name[0] > '9')) {
+				rpt_telemetry(myrpt, CONNECTED, l);
+			}
+		} else {
+			rpt_mutex_unlock(&myrpt->lock);
+		}
+
+		/* In theory, both these conditions should be true if one is, since the node thread will queue a soft hangup on this channel and then set l->chan to NULL */
+		if (!l->chan || ast_check_hangup(chan)) {
```

Comment:
I'm thinking we need to set `newkeytimer = 0;` here since we are no longer looking for a newkey message.

Reply:
Let's deal with this later if needed.